DataAccessUtils result accessors with Optional return value

KaidosGH opened the pull request:
Closes #27728
(https://github.com/spring-projects/spring-framework/issues/27728)

Hi everyone, this is my first PR in any project here on github with the goal of improving my understanding of java. Hope i understood the enhancement request correctly.

KaidosGH committed:
extend DataAccessUtils

KaidosGH committed:
cleanup

BrucessKING commented:
Using the '.*' form of import should be avoided - java.util.*.

marschall commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +61,109 @@
+			return null;
+		}
+		Iterable<T> iterable = () -> results;
+		List<T> resultList = StreamSupport.stream(iterable.spliterator(), false).toList();
```
Why a stream instead of just

    results.next()

KaidosGH committed:
optimize imports, refactor iterator usage

KaidosGH commented:
> Using the * form of import should be avoided - java.util.*.

Thanks for the tip, i adjusted that.

KaidosGH commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +61,109 @@
+		if (results == null) {
+			return null;
+		}
+		List<T> resultList = results.toList();
```
Hi @marschall,

thanks for reviewing, sounds good, i refactored this. Does it make more sense now?

`throw new IncorrectResultSizeDataAccessException(1, resultList.size());`
as we limited the resultList size with .limit, this would not report the original passed Stream results size though but instead a maximum of 2... is that an issue?

marschall replied:
Yes, looks better to me.

marschall commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +60,107 @@
+		if (results == null) {
+			return null;
+		}
+		List<T> resultList = results.limit(2).toList();
```
There is a trade off here and in `#optionalResult` whether we should call `#close` on the stream:

On one hand Java convention is the code who creates a "resource" is responsible for disposing it, in this case the caller of this method would be responsible to call `#close` on the streams where required.
On the other hand failing to call `#close` on some streams, like the ones returned by `JdbcOperations#queryForStream`, can lead to resource leaks. Having the caller put streams in a try-with-resources block is both cumbersome and error prone.

KaidosGH replied:
Makes sense. The passed stream is now closed in both methods. Also e.g. hibernate's Query.stream() documentation suggests the stream should be closed afterwards, so not so far fetched 👍

KaidosGH committed:
close streams properly with try-with-resource

rkvigneswaran commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +60,111 @@
+	 * element has been found in the given Collection
+	 */
+	public static <T> Optional<T> optionalResult(@Nullable Collection<T> results) throws IncorrectResultSizeDataAccessException {
+		if (CollectionUtils.isEmpty(results)) {
+			return Optional.empty();
+		}
+		if (results.size() > 1) {
+			throw new IncorrectResultSizeDataAccessException(1, results.size());
+		}
+		return results.stream().findFirst();
```
Would it be better if we delegate this to original `singleResult()` method

```suggestion
		return Optional.ofNullable(singleResult(results));
```

KaidosGH replied:
Hi @rkvigneswaran,

that is actually pretty smart. Cuts out a lot of redundant code! 👍

rkvigneswaran commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +60,111 @@
+	 * element has been found in the given Stream
+	 */
+	public static <T> Optional<T> optionalResult(@Nullable Stream<T> results) throws IncorrectResultSizeDataAccessException {
+		if (results == null) {
+			return Optional.empty();
+		}
+		try (results) {
+			List<T> resultList = results.limit(2).toList();
+			if (resultList.size() > 1) {
+				throw new IncorrectResultSizeDataAccessException(1, resultList.size());
+			}
+			return CollectionUtils.isEmpty(resultList) ? Optional.empty() : Optional.of(resultList.get(0));
```
Would it be better if we delegate this to newly added singleResult(Stream<T>) method?
```suggestion
		return Optional.ofNullable(singleResult(results));
```

rkvigneswaran commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +60,111 @@
+	 * element has been found in the given Iterator
+	 */
+	public static <T> Optional<T> optionalResult(@Nullable Iterator<T> results) throws IncorrectResultSizeDataAccessException {
+		if (results == null) {
+			return Optional.empty();
+		}
+		T result = results.hasNext() ? results.next() : null;
+		if (results.hasNext()) {
+			throw new IncorrectResultSizeDataAccessException(1, 2);
+		}
+		return Optional.ofNullable(result);
```
Would it be better if we delegate this to newly added `singleResult(Iterator<T>)` method?
```suggestion
		return Optional.ofNullable(singleResult(results));
```

rkvigneswaran reviewed (changes requested):
Hi,

This is my first review comments in any open source project. Kindly let me know if my comments are invalid.

KaidosGH committed:
Update spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

re-use newly added singleResult methods

Co-authored-by: Vigneswaran <rkvigneswaran@gmail.com>

KaidosGH committed:
Update spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

re-use newly added singleResult methods

Co-authored-by: Vigneswaran <rkvigneswaran@gmail.com>

KaidosGH committed:
Update spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

re-use newly added singleResult methods

Co-authored-by: Vigneswaran <rkvigneswaran@gmail.com>

KaidosGH committed:
fix closing bracket

bananayong commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +60,89 @@
+		try (results) {
+			List<T> resultList = results.limit(2).toList();
+			if (resultList.size() > 1) {
+				throw new IncorrectResultSizeDataAccessException(1, resultList.size());
```
I think IncorrectResultSizeDataAccessException should contain the size of the results stream.
if `resultList.size()` will be used, actual size of IncorrectResultSizeDataAccessException is `<=2` always unlike actual size of results.

If prefer stream way, How about utilze `teeing`?
```java
		results.collect(teeing(counting(), reducing((first, second) -> first), (count, result) -> {
			if (count > 1) {
				throw new IncorrectResultSizeDataAccessException(1, count.intValue());
			}
			return result.orElse(null);
		}))
```

rkvigneswaran replied:
Just a suggestion. There are multiple constructors provided by `IncorrectResultSizeDataAccessException` class. 

* One of the constructors doesn't take `actualSize` argument and assigns the value `-1` if the size if unknown. You can consider using it
  * `public IncorrectResultSizeDataAccessException(int expectedSize)` - [Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/dao/IncorrectResultSizeDataAccessException.html#IncorrectResultSizeDataAccessException-int-)
* Another one explicitly suggests setting the value to `-1` if size is unknown
  * `public IncorrectResultSizeDataAccessException(int expectedSize, int actualSize)` - [Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/dao/IncorrectResultSizeDataAccessException.html#IncorrectResultSizeDataAccessException-int-int-)

Both of them would be suitable for this scenario since size is unknown

bananayong replied:
It makes sense.

rkvigneswaran replied:
```suggestion
				throw new IncorrectResultSizeDataAccessException(1);
```

Adding as suggestion for convenience

KaidosGH replied:
yeah let's use the `public IncorrectResultSizeDataAccessException(int expectedSize)` for the iterator and stream method. 
What do you mean by "UT"?

rkvigneswaran commented on spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java:
```diff
@@ -56,6 +60,89 @@
+		}
+		T result = results.hasNext() ? results.next() : null;
+		if (results.hasNext()) {
+			throw new IncorrectResultSizeDataAccessException(1, 2);
```
Same suggestion as above : [Link](https://github.com/spring-projects/spring-framework/pull/27735/files#r757849057)
```suggestion
			throw new IncorrectResultSizeDataAccessException(1);
```
Since `actualSize` of the iterator isn't known beforehand, we can use `public IncorrectResultSizeDataAccessException(int expectedSize)` constructor and set the `actualSize`  as `-1`

rkvigneswaran reviewed (changes requested):
Modifying the exception to make it clear that size of the underlying Stream/Iterator is not known

KaidosGH committed:
Update spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

use "public IncorrectResultSizeDataAccessException(int expectedSize)" in stream and iterator method

Co-authored-by: Kwangyong Kim <banana.yong@gmail.com>

KaidosGH committed:
Update spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

use "public IncorrectResultSizeDataAccessException(int expectedSize)" in stream and iterator method

Co-authored-by: Vigneswaran <rkvigneswaran@gmail.com>

KaidosGH committed:
update tests to correctly assert the thrown exception
